pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,723 @@
+			} else {
+				logrus.Infof("waiting for cluster create [%s] to start", cluster.Name)
+			}
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster AKS [%s] pre-create failure to be resolved", cluster.Name)
+		return e.setFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	}
+}
+
+func (e *aksOperatorController) setInitialUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	logrus.Infof("setting initial upstreamSpec on cluster [%s]", cluster.Name)
+	cluster = cluster.DeepCopy()
+	upstreamSpec, err := clusterupstreamrefresher.BuildAKSUpstreamSpec(e.secretsCache, cluster)
+	if err != nil {
+		return cluster, err
+	}
+	cluster.Status.AKSStatus.UpstreamSpec = upstreamSpec
+	return e.clusterClient.Update(cluster)
+}
+
+// updateAKSClusterConfig updates the AKSClusterConfig object's spec with the cluster's AKSConfig if they are not equal..
+func (e *aksOperatorController) updateAKSClusterConfig(cluster *mgmtv3.Cluster, aksClusterConfigDynamic *unstructured.Unstructured, spec map[string]interface{}) (*mgmtv3.Cluster, error) {
+	list, err := e.dynamicClient.Namespace(namespace.GlobalNamespace).List(context.TODO(), v1.ListOptions{})
+	if err != nil {
+		return cluster, err
+	}
+	selector := fields.OneTermEqualSelector("metadata.name", cluster.Name)
+	w, err := e.dynamicClient.Namespace(namespace.GlobalNamespace).Watch(context.TODO(), v1.ListOptions{ResourceVersion: list.GetResourceVersion(), FieldSelector: selector.String()})
+	if err != nil {
+		return cluster, err
+	}
+	aksClusterConfigDynamic.Object["spec"] = spec
+	aksClusterConfigDynamic, err = e.dynamicClient.Namespace(namespace.GlobalNamespace).Update(context.TODO(), aksClusterConfigDynamic, v1.UpdateOptions{})
+	if err != nil {
+		return cluster, err
+	}
+
+	// AKS cluster and node pool statuses are not always immediately updated. This cause the AKSConfig to
+	// stay in "active" for a few seconds, causing the cluster to go back to "active".
+	timeout := time.NewTimer(10 * time.Second)
+	for {
+		select {
+		case event := <-w.ResultChan():
+			aksClusterConfigDynamic = event.Object.(*unstructured.Unstructured)
+			status, _ := aksClusterConfigDynamic.Object["status"].(map[string]interface{})
+			if status["phase"] == "active" {
+				continue
+			}
+
+			// this enqueue is necessary to ensure that the controller is reentered with the updating phase
+			e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// recordCAAndAPIEndpoint reads the AKSClusterConfig's secret once available. The CA cert and API endpoint are then copied to the cluster status.
+func (e *aksOperatorController) recordCAAndAPIEndpoint(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
```

Comment:
This function looks very similar to the one in the EKS integration controller, if we continue creating a secret of the same format in the AKS operator can we just have EKS and AKS controllers use the same logic for this? I don't think anything in here is provider specific or uses anything from one of the config fields.

Reply:
Ok, it will be reworked

Reply:
We now have `clusteroperator/utils.go` for this